iec_api/data.py

```diff
@@ -0,0 +1,227 @@
+import json
+import logging
+
+import requests
+
+from iec_api.commons import add_jwt_to_headers
+from iec_api.const import (
+    GET_BILLING_INVOICES,
+    GET_CONSUMER_URL,
+    GET_CONTRACTS_URL,
+    GET_DEFAULT_CONTRACT_URL,
+    GET_DEVICE_TYPE_URL,
+    GET_DEVICES_BY_CONTRACT_ID_URL,
+    GET_DEVICES_URL,
+    GET_ELECTRIC_BILL_URL,
+    GET_LAST_METER_READING_URL,
+    GET_REQUEST_READING_URL,
+    HEADERS_WITH_AUTH,
+)
+from iec_api.models.contract import GetContractResponse
+from iec_api.models.customer import Customer
+from iec_api.models.device import Device, Devices, GetDeviceResponse
+from iec_api.models.device_type import DeviceType
+from iec_api.models.electric_bill import GetElectricBillResponse
+from iec_api.models.exceptions import IECError
+from iec_api.models.invoice import GetInvoicesBody, GetInvoicesResponse
+from iec_api.models.jwt import JWT
+from iec_api.models.meter_reading import GetLastMeterReadingResponse
+from iec_api.models.remote_reading import RemoteReadingRequest, RemoteReadingResponse
+from iec_api.models.response_descriptor import ErrorResponseDescriptor
+
+logger = logging.getLogger(__name__)
+
+
+def _get_url(url, headers):
+    logger.debug("HTTP GET: %s", url)
+    return requests.get(url=url, headers=headers, timeout=10)
+
+
+def get_customer(token: JWT) -> Customer:
```

Comment:
All of your functions are structured like:

```python
def get_something(token: JWT) -> Something:
    headers = add_jwt_to_headers(HEADERS_WITH_AUTH, token.id_token)
    construct_request()
    get_response()
    handle_bad_response()
    return_deserialized_object()
```

I think that you can use a sort of "factory function" to auto-generate all of these functions.

Something like that:

```python
from typing import Any, Callable
...
def get_iec_api_getter(response_class: Any) -> Callable:
    def func(token: JWT, url: str, *args, **kwargs) -> Any:
        headers = add_jwt_to_headers(HEADERS_WITH_AUTH, token.id_token)
        url = url.format(kwargs) if kwargs else url
        response = _get_url(url=url, headers=headers)
        if response.status_code != http.HTTPStatus.OK:  # 200
            ...
        return response_class.from_dict(response.json())
    return func
```

And then:
```python
get_customer = get_iec_api_getter(response_class=Customer)
get_electric_bill = get_iec_api_getter(response_class=GetElectricBillResponse)
```

The problems with this method as I see it:
1. Will not work for the remote reading function (can be altered to work)
2. Not as explicit, will not show proper documentation in tools like iPython, thus harder to debug

Reply:
As mentioned personally - I'm moving this whole project to async - and did just this:
https://github.com/GuyKh/py-iec-api/blob/1a734209df37d755ac6825d16b9041b0dac23442/iec_api/data.py#L48C1-L71C41